src/messaging/ExchangeHandle.h

```diff
@@ -0,0 +1,62 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+namespace chip {
+
+namespace Messaging {
+
+class ExchangeContext;
+
+/**
+ * ExchangeHandle acts like a unique_ptr to an ExchangeContext, to help managing the life cycle of the ExchangeContext
+ */
+class ExchangeHandle
+{
+public:
+    explicit ExchangeHandle() : mContext(nullptr) {}
+    explicit ExchangeHandle(ExchangeContext * context);
+    ~ExchangeHandle();
+
+    ExchangeHandle(const ExchangeHandle & that);
+    ExchangeHandle(ExchangeHandle && that) : mContext(that.mContext) { that.mContext = nullptr; }
+
+    ExchangeHandle & operator=(const ExchangeHandle & that);
+    ExchangeHandle & operator=(ExchangeHandle && that)
```

Comment:
Why is this function inlined, unlike all the other ones?

Reply:
This one doesn's depend on the implementation of `ExchangeContext`, but others depend on it.

As here `ExchangeContext` is a forward declaration, others can be not implemented here.

Reply:
Right, but why do we want to inline this one at all?

Reply:
I have added a template class `ReferenceCountedHandle` for this part. instead of `ExchangeHandle`